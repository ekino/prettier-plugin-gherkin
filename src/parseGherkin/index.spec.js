const fs = require("fs");
const path = require("path");
const parseGherkin = require(".");

describe("parseGherkin", () => {
  it("parse basic feature file into an AST", () => {
    const basicFeature = fs.readFileSync(
      path.join(__dirname, "__fixtures__/basic.feature"),
      { encoding: "utf-8" },
    );

    const ast = parseGherkin(basicFeature);

    expect(ast).toMatchSnapshot();
  });

  it("parse feature file with data tables into an AST", () => {
    const basicFeature = fs.readFileSync(
      path.join(__dirname, "__fixtures__/with_data_tables.feature"),
      { encoding: "utf-8" },
    );

    const ast = parseGherkin(basicFeature);

    expect(ast).toMatchSnapshot();
  });

  it("should throw an exception when the is a syntax error", () => {
    const fixture = `Scenario: foo`; // No "Feature:" header

    expect(() => parseGherkin(fixture)).toThrow(
      "(1:1): expected: #EOF, #Language, #TagLine, #FeatureLine, #Comment, #Empty, got 'Scenario: foo'",
    );
  });

  it("should parse tags on Feature", () => {
    const fixture = `@foo @bar
Feature: baz`;

    const ast = parseGherkin(fixture);

    expect(ast).toMatchSnapshot();
  });

  it("should parse tags on Scenario", () => {
    const fixture = `Feature: some feature
  @foo @bar
  Scenario: some scenario`;

    const ast = parseGherkin(fixture);

    expect(ast).toMatchSnapshot();
  });

  it("should parse comments", () => {
    const fixture = `# Some comment
Feature: Lorem ipsum dolor sit amet

  # Another one
  Scenario: foo
`;

    const ast = parseGherkin(fixture);

    expect(ast).toMatchSnapshot();
  });
});
