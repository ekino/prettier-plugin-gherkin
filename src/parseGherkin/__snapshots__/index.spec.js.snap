// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseGherkin parse basic feature file into an AST 1`] = `
Array [
  Object {
    "description": null,
    "keyword": "Feature",
    "language": "en",
    "location": Object {
      "column": 1,
      "line": 1,
    },
    "name": "Minimal",
    "tags": Array [],
    "type": "feature",
  },
  Object {
    "description": null,
    "keyword": "Scenario",
    "location": Object {
      "column": 5,
      "line": 2,
    },
    "name": "minimalistic",
    "tags": Array [],
    "type": "scenario",
  },
  Object {
    "keyword": "Given ",
    "location": Object {
      "column": 9,
      "line": 3,
    },
    "text": "the minimalism",
    "type": "step",
  },
  Object {
    "keyword": "When ",
    "location": Object {
      "column": 9,
      "line": 4,
    },
    "text": "you change nothing",
    "type": "step",
  },
  Object {
    "keyword": "Then ",
    "location": Object {
      "column": 9,
      "line": 5,
    },
    "text": "it's still small",
    "type": "step",
  },
]
`;

exports[`parseGherkin parse feature file with data tables into an AST 1`] = `
Array [
  Object {
    "description": "  As an API client I can request my api",
    "keyword": "Feature",
    "language": "en",
    "location": Object {
      "column": 1,
      "line": 2,
    },
    "name": "Title of feature",
    "tags": Array [
      Object {
        "location": Object {
          "column": 1,
          "line": 1,
        },
        "name": "@tag",
      },
    ],
    "type": "feature",
  },
  Object {
    "location": Object {
      "column": 1,
      "line": 6,
    },
    "text": "  #Comment",
    "type": "comment",
  },
  Object {
    "description": null,
    "keyword": "Scenario",
    "location": Object {
      "column": 3,
      "line": 7,
    },
    "name": "Query on API, with success response",
    "tags": Array [],
    "type": "scenario",
  },
  Object {
    "keyword": "Given ",
    "location": Object {
      "column": 5,
      "line": 9,
    },
    "text": "I load test data files",
    "type": "step",
  },
  Object {
    "location": Object {
      "column": 7,
      "line": 10,
    },
    "rows": Array [
      Object {
        "cells": Array [
          Object {
            "location": Object {
              "column": 9,
              "line": 10,
            },
            "maxLength": 22,
            "type": "cell",
            "value": "path",
          },
          Object {
            "location": Object {
              "column": 34,
              "line": 10,
            },
            "maxLength": 5,
            "type": "cell",
            "value": "type",
          },
        ],
        "location": Object {
          "column": 7,
          "line": 10,
        },
        "type": "row",
      },
      Object {
        "cells": Array [
          Object {
            "location": Object {
              "column": 9,
              "line": 11,
            },
            "maxLength": 22,
            "type": "cell",
            "value": "data/mocks/data.json",
          },
          Object {
            "location": Object {
              "column": 34,
              "line": 11,
            },
            "maxLength": 5,
            "type": "cell",
            "value": "mocks",
          },
        ],
        "location": Object {
          "column": 7,
          "line": 11,
        },
        "type": "row",
      },
      Object {
        "cells": Array [
          Object {
            "location": Object {
              "column": 9,
              "line": 12,
            },
            "maxLength": 22,
            "type": "cell",
            "value": "data/redis/tokens.json",
          },
          Object {
            "location": Object {
              "column": 34,
              "line": 12,
            },
            "maxLength": 5,
            "type": "cell",
            "value": "redis",
          },
        ],
        "location": Object {
          "column": 7,
          "line": 12,
        },
        "type": "row",
      },
    ],
    "type": "dataTable",
  },
  Object {
    "keyword": "When ",
    "location": Object {
      "column": 5,
      "line": 13,
    },
    "text": "I GET /api/v2/route",
    "type": "step",
  },
]
`;

exports[`parseGherkin should parse comments 1`] = `
Array [
  Object {
    "location": Object {
      "column": 1,
      "line": 1,
    },
    "text": "# Some comment",
    "type": "comment",
  },
  Object {
    "description": null,
    "keyword": "Feature",
    "language": "en",
    "location": Object {
      "column": 1,
      "line": 2,
    },
    "name": "Lorem ipsum dolor sit amet",
    "tags": Array [],
    "type": "feature",
  },
  Object {
    "location": Object {
      "column": 1,
      "line": 4,
    },
    "text": "  # Another one",
    "type": "comment",
  },
  Object {
    "description": null,
    "keyword": "Scenario",
    "location": Object {
      "column": 3,
      "line": 5,
    },
    "name": "foo",
    "tags": Array [],
    "type": "scenario",
  },
]
`;

exports[`parseGherkin should parse tags on Feature 1`] = `
Array [
  Object {
    "description": null,
    "keyword": "Feature",
    "language": "en",
    "location": Object {
      "column": 1,
      "line": 2,
    },
    "name": "baz",
    "tags": Array [
      Object {
        "location": Object {
          "column": 1,
          "line": 1,
        },
        "name": "@foo",
      },
      Object {
        "location": Object {
          "column": 6,
          "line": 1,
        },
        "name": "@bar",
      },
    ],
    "type": "feature",
  },
]
`;

exports[`parseGherkin should parse tags on Scenario 1`] = `
Array [
  Object {
    "description": null,
    "keyword": "Feature",
    "language": "en",
    "location": Object {
      "column": 1,
      "line": 1,
    },
    "name": "some feature",
    "tags": Array [],
    "type": "feature",
  },
  Object {
    "description": null,
    "keyword": "Scenario",
    "location": Object {
      "column": 3,
      "line": 3,
    },
    "name": "some scenario",
    "tags": Array [
      Object {
        "location": Object {
          "column": 3,
          "line": 2,
        },
        "name": "@foo",
      },
      Object {
        "location": Object {
          "column": 8,
          "line": 2,
        },
        "name": "@bar",
      },
    ],
    "type": "scenario",
  },
]
`;
